### PR TITLE
Ability to export pixel_unshuffle to ONNX

### LINKED ISSUE #51181: Ability to export `pixel_unshuffle` to ONNX
## 🚀 Feature
Request support for exporting the new `nn.PixelUnshuffle` / `F.pixel_unshuffle` operation to ONNX using `torch.onnx.export`.

## Motivation
The "PixelUnshuffle" operation, also known as "SpaceToDepth", was recently added to torch with [this feature request](https://github.com/pytorch/pytorch/issues/2456). This operation is the inverse of the "PixelShuffle" / "DepthToSpace" operation, and both operations have been [supported by ONNX](https://github.com/onnx/onnx/blob/master/docs/Operators.md) since at least opset 9. However, torch only supports exporting PixelShuffle, and *not* Pixel**Un**shuffle, to ONNX. This precludes users' ability to deploy models trained in torch that include PixelUnshuffle layers to ONNX (and other) runtimes. Such models include convolutional U-nets, autoencoders, super-resolution models, and many others that require efficient lossless spatial re-sampling. 

Example:
```python
import torch

layer = torch.nn.PixelUnshuffle(2)

x = torch.rand(1, 1, 2, 2)

assert (layer(x) == x.view(1, 4, 1, 1)).all().item()  # passes

for opset_version in range(9, 14):
    try:
        torch.onnx.export(layer, x, "./pixel_unshuffle.onnx", opset_version=opset_version)
    except RuntimeError as err:
        print(err)
```
Prints for every version:
```
Exporting the operator pixel_unshuffle to ONNX opset version <version> is not supported. Please open a bug to request ONNX export support for the missing operator.
```

## Pitch
In order to achieve parity with PixelShuffle, and in order to support deploying the many vision models that require it, I request that torch developers please add support for exporting the PixelUnshuffle operation to ONNX.

## Alternatives
This feature is in fact present in tensorflow and tf2onnx, so I am currently using that framework for my particular use case. This induces hassle and stress among my group, as we would prefer to stick to a single framework.

## Additional context
I was advised to file a feature request by @mruberry [here](https://github.com/pytorch/pytorch/issues/2456#issuecomment-759835306).

cc @houseroad @spandantiwari @lara-hdr @BowenBao @neginraoof @gmagogsfm

### PR DESCRIPTION
**Pull request by jkyl**
Fixes #51181 by registering torch's `pixel_unshuffle` op with ONNX's `SpaceToDepth` op. Also adds 2 ONNX tests that mirror existing tests for `pixel_shuffle` (the inverse op).

### CODE REVIEW

File: torch/onnx/symbolic_opset11.py [hunk 1]
```diff
@@ -207,6 +207,14 @@ def pixel_shuffle(g, self, upscale_factor):
     return g.op("DepthToSpace", self, blocksize_i=upscale_factor, mode_s="CRD")
 
 
+@parse_args('v', 'i')
+def pixel_unshuffle(g, self, downscale_factor):
```
Comment (shubhambhokare1) [inline torch/onnx/symbolic_opset11.py:211]: Could we add a symbolic in opset 9 to support more opsets?

File: test/onnx/test_operators.py [hunk 2]
```diff
@@ -814,6 +814,10 @@ def test_pixel_shuffle(self):
         x = torch.randn(2, 8, 3, 4).float()
         self.assertONNX(lambda x: torch.pixel_shuffle(x, upscale_factor=2), x, opset_version=11)
 
+    def test_pixel_unshuffle(self):
```
Comment (shubhambhokare1) [inline test/onnx/test_operators.py:817]: This test is failing, every time you add tests to test_operators run python test/onnx/test_operators.py with --accept parameter. See below for more details
```
Usage: python test/onnx/test_operators.py [--no-onnx] [--produce-onnx-test-data]
          --no-onnx: no onnx python dependence
          --produce-onnx-test-data: generate onnx test data
          --accept: accept onnx updates and overwrite models
```

**Comment by haydn-jones** [comment]
Tried this out on the nightly builds of PyTorch with my super resolution model and saw that colors were getting screwed up so I looked into it. Note that I simply copied and pasted the implementation from `symbolic_opset11.py` from this PR into my local install.

As far as I can tell, ONNX is doing the unshuffling order differently than PyTorch. Code is below to see what I mean. Interestingly enough, the DepthToSpace operation has a `mode_s` argument to control the ordering, but SpaceToDepth does not.

```Python
import onnxruntime
import torch
import torch.nn as nn

torch.set_grad_enabled(False)

image = torch.arange(0, 2*16).float().reshape(1, 2, 4, 4)

model = nn.PixelUnshuffle(2)

torch_out = model(image)

torch.onnx.export(model,
    image,
    "test.onnx",
    opset_version=11,
    input_names =['input'],
    output_names=['output'],
)

ort_session = onnxruntime.InferenceSession("test.onnx")

ort_inputs = {ort_session.get_inputs()[0].name: image.numpy()}
ort_out = torch.from_numpy(ort_session.run(None, ort_inputs)[0]).squeeze()

print(f"Torch:\n", torch_out.squeeze())
print(f"ONNX:\n", ort_out.squeeze())
print(f"Diff:\n", (torch_out-ort_out).squeeze())
```
A side by side printout of the tensors is as follows
image

Earlier I ended up implementing PixelUnshuffle myself to allow ONNX export and it works fine. Feel free to use it in place of the PyTorch PixelUnshuffle if you need your model exported (or simply replace the unshuffle layers during export, which is what I'm doing). Unfortunately it creates a comically ugly graph for this operation, but it works nonetheless.
```Python
class PixelUnshuffle(nn.Module):
    def __init__(self, scale_factor: int) -> None:
        super().__init__()

        self.scale_factor = scale_factor

    def forward(self, x: torch.Tensor) -> torch.Tensor:
        b, c, h, w = x.size()

        out_channel = c*(self.scale_factor**2)

        out_h = h//self.scale_factor
        out_w = w//self.scale_factor

        x = x.view(b, c, out_h, self.scale_factor, out_w, self.scale_factor)
        
        return x.permute(0, 1, 3, 5, 2, 4).reshape(b,out_channel, out_h, out_w)
```

**Comment by shubhambhokare1** [comment]
PR #72449 should add pixel_unshuffle support. We will have to wait for ONNX's mode_s implementation in SpaceToDepth operator until we can add an updated implementation.

**Comment by justinchuby** [comment]
Closed because pixel_unshuffle support is added. Please reopen if you think this was a mistake. Thanks for your contribution!
